modules/drivers/mongo/src/metabase/driver/mongo/util.clj

```diff
@@ -89,6 +88,64 @@
     :else                         (throw (Exception. (str "with-mongo-connection failed: bad connection details:"
                                                           (:details database))))))
 
+(defn- connect-via-uri
+  [^MongoClientURI uri ^MongoClient conn]
+  (if-let [dbName (.getDatabase uri)]
+    (.getDB conn dbName)
+    (throw (IllegalArgumentException. "No database name specified in URI. Monger requires a database to be explicitly configured."))))
+
+(defn- connect
+  "Connects to Mongo.  This is the fallback method to connect to hostnames that are
+   not FQDNs.  This works with 'localhost', but has been problematic with FQDNs.
+   If you would like to provide a FQDN, use `connect-srv`."
+  [{:keys [host port user authdb pass dbname ssl additional-options]}]
+  (let [server-address (mg/server-address host port)
+        credentials (when user
+                      (mcred/create user authdb pass))
+        ^MongoClientOptions$Builder opts (connection-options-builder :ssl? ssl, :additional-options additional-options)
+        do-connect (partial mg/connect server-address (-> opts .build))
+        mongo-client (if credentials
+                       (do-connect credentials)
+                       (do-connect))
+        db (mg/get-db mongo-client dbname)]
+    [mongo-client db]))
+
+(defn- connect-srv
+  "Connects to Mongo using DNS SRV.  Requires FQDN for `host` in the format
+   'hostname.domain.top-level-domain'.  Only a single host is supported, but a
+   replica list could easily provided instead of a single host.
+
+   Using SRV automatically enables SSL, though we explicitly set SSL to true anyway.
+
+   Docs to generate URI string:
+   https://docs.mongodb.com/manual/reference/connection-string/#dns-seedlist-connection-format"
+  [{:keys [host port user authdb pass dbname ssl additional-options]}]
+  (let [conn-opts (connection-options-builder :ssl? ssl, :additional-options additional-options)
+        authdb (if (seq authdb)
+                 authdb
+                 dbname)
+        conn-str (format "mongodb+srv://%s:%s@%s/%s" user pass host authdb)
+        mongo-uri (MongoClientURI. conn-str conn-opts)
+        mongo-client (MongoClient. mongo-uri)
+        db (connect-via-uri mongo-uri mongo-client)]
+    [mongo-client db]))
+
+(defn- fqdn?
+  "A very simple way to check if a hostname is fully-qualified:
+   Check if there are exactly two periods in the name."
+  [host]
+  (= 2 (-> host frequencies (get \.))))
+
+(defn- connect-fn
+  "If `host` is a fully-qualified domain name, then we need to connect to Mongo
+   differently.  It has been problematic to connect to Mongo with an FQDN using
+   `mg/connect`.  The fix was to create a connection string and use DNS SRV for
+   FQDNS.  In this fn we provide the correct connection fn based on host."
+  [host]
+  (if (fqdn? host)
+    connect-srv
+    connect))
+
 (defn -with-mongo-connection
```

Comment:
I wrote this Mongo code like 5 years ago when I was still relatively new to writing Clojure full-time. So if you want to refactor it to make it clearer or better please do not hesitate to do so. Our general rule if you want to make something clearer or better do it.

I was thinking about how one would write a test for this stuff and I think it would really be helpful to break this out into a few smaller functions. Specifically, I would decouple the code that generates a MongoDB connection object with the code that actually uses that object to establish a connection.

I played with this for a minute and think something like this will serve us better: it will be 100% easier to test, easier to follow (IMO) and extensible if we ever need to support one of the other options as well (hopefully not):

https://gist.github.com/camsaul/81bc81338726d4e618b8a36ce7433ff9

Reply:
I guess we can't comment on a Gist but my thinking was move the code that takes the Database `details` map and converts that to the appropriate MongoDB function into a couple of functions:

*  `normalize-details` to add defaults, remove empty keys like `:pass` (where needed) and rename keys stuff as appropriate
*  `details->mongo-connection-info` which will take the details and then return a map with a `:type` key for the type of connection we will be using. Other stuff such as URI are included in the map.

The we can have a separate `connect` function that takes the `mongo-connection-info` map and actually makes the connection. We can make it a multimethod that dispatches off of `:type` because both implementations are supposed to be returning the same thing.

The new tests will be super simple, you can just test `details->mongo-connection-info` directly and make sure that given the correct input from Database `:details` we get the correct connection object types

Reply:
Those are just suggestions, I'm just trying to think of ways to make your life easier writing the tests

Reply:
Nice, good suggestion, I'll re-organize